File: app/src/main/java/com/aspire/dubsmash/siavash/AdapterSounds.java
package com.aspire.dubsmash.siavash;

import android.app.Activity;
import android.app.ProgressDialog;
import android.content.Context;
import android.content.Intent;
import android.media.MediaPlayer;
import android.net.Uri;
import android.os.AsyncTask;
import android.support.annotation.MainThread;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.ProgressBar;
import android.widget.RelativeLayout;
import android.widget.TextView;
import android.widget.Toast;

import com.aspire.dubsmash.R;
import com.aspire.dubsmash.hojjat.ActivityRecordDub;
import com.aspire.dubsmash.util.Constants;
import com.aspire.dubsmash.util.PersianReshape;
import com.aspire.dubsmash.util.Util;

import java.io.File;
import java.io.IOException;
import java.util.List;

import butterknife.Bind;
import butterknife.ButterKnife;
import butterknife.OnClick;
import retrofit.Callback;
import retrofit.RetrofitError;
import retrofit.client.Response;

/**
 * Created by sia on 9/30/15.
 */
public class AdapterSounds extends RecyclerView.Adapter<AdapterSounds.SoundsViewHolder> {
    private static final String TAG = AdapterSounds.class.getSimpleName();
    private List<Sound> mSounds;
    private boolean isPlaying;
    private ImageButton mPlayingSoundButton;
    private MediaPlayer mMediaPlayer;

    public AdapterSounds(List<Sound> sounds) {
        mSounds = sounds;
    }

    @Override
    public SoundsViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_sounds, parent, false);
        return new SoundsViewHolder(view);
    }

    @Override
    public void onBindViewHolder(SoundsViewHolder holder, int position) {
        holder.bindViews(mSounds.get(position));
    }

    @Override
    public int getItemCount() {
        return mSounds.isEmpty() ? 1 : mSounds.size();
    }

    public void addData(List<Sound> data) {
        mSounds = data;
        Log.d(TAG, mSounds.size() + "");
        notifyDataSetChanged();
    }

    public class SoundsViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
        @Bind(R.id.item) RelativeLayout mItem;
        @Bind(R.id.title) TextView mTitle;
        @Bind(R.id.uploaded_by) TextView mUploadedBy;
        @Bind(R.id.like) ImageButton mLikeButton;
        @Bind(R.id.play_sound) ImageButton mPlaySound;

        private Context mContext;
        private Sound mSound;
        private boolean isLiked;

        public SoundsViewHolder(View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);

            mContext = itemView.getContext();
            isLiked = false;
            setListeners();
        }

        private void bindViews(Sound sound) {
            mTitle.setText(sound.getSoundTitle());
            mUploadedBy.setText(sound.getSoundSendername());
            mSound = sound;
        }

        private void setListeners() {
            mItem.setOnClickListener(this);
            mLikeButton.setOnClickListener(this);
            mPlaySound.setOnClickListener(this);
        }

        @OnClick(R.id.go_to_dubmaker) void goToDubMaker() {

        }

        @Override public void onClick(View v) {
            int id = v.getId();
            if (id == mLikeButton.getId()) {
                if (!isLiked) {
                    mLikeButton.setImageResource(R.drawable.favorite_selected);
                    ActivityMain.sNetworkApi.getSoundLikes(Util.getUserId(mContext), mSound.getSoundId(), new Callback<Response>() {
                        @Override public void success(Response response, Response response2) {

                        }

                        @Override public void failure(RetrofitError error) {

                        }
                    });
                    isLiked = true;
                } else {
                    mLikeButton.setImageResource(R.drawable.favorite_unselected);
                    isLiked = false;
                }
            } else if (id == mPlaySound.getId()) {
                if (isPlaying) {
                    if (mMediaPlayer != null) {
                        if (mMediaPlayer.isPlaying())
                            mMediaPlayer.stop();
                        mMediaPlayer.reset();
                        mMediaPlayer.release();
                        mMediaPlayer = null;
                    }
                    mPlayingSoundButton.setImageResource(R.drawable.ic_play);
                    mPlayingSoundButton = mPlaySound;
                    mPlaySound.setImageResource(R.drawable.download);
                } else {
                    mPlayingSoundButton = mPlaySound;
                    mPlaySound.setImageResource(R.drawable.download);
                    isPlaying = true;
                }
                ActivityMain.sNetworkApi.downloadSingleData(mSound.getSoundUrl(), new Callback<Response>() {
                    @Override public void success(Response response, Response response2) {
                        try {
                            Util.saveToFile(Util.getFromStream(response.getBody().in()), new File(Constants.TEMP_SOUND_PATH).getPath());
                            if (mPlayingSoundButton == mPlaySound) {
                                mPlaySound.setImageResource(R.drawable.ic_pause);
                                mMediaPlayer = MediaPlayer.create(mContext, Uri.parse(Constants.TEMP_SOUND_PATH));
                                mMediaPlayer.start();
                                mMediaPlayer.setOnCompletionListener(new MediaPlayer.OnCompletionListener() {
                                    @Override public void onCompletion(MediaPlayer mp) {
                                        mPlaySound.setImageResource(R.drawable.ic_play);
                                        mMediaPlayer.reset();
                                        mMediaPlayer.release();
                                        mMediaPlayer = null;
                                        isPlaying = false;
                                    }
                                });
                            }
                        } catch (IOException e) {
                            e.printStackTrace();
                        }
                    }

                    @Override public void failure(RetrofitError error) {

                    }
                });
            } else if (id == mItem.getId()) {
                final ProgressDialog progressDialog = new ProgressDialog(mContext);
                progressDialog.setMessage(PersianReshape.reshape("کمی صبر کنید"));
                progressDialog.setCancelable(false);
                progressDialog.show();
                ActivityMain.sNetworkApi.downloadSingleData(mSound.getSoundUrl(), new Callback<Response>() {
                    @Override
                    public void success(Response response, Response response2) {
                        progressDialog.dismiss();
                        try {
                            Util.saveToFile(Util.getFromStream(response.getBody().in()), new File(Constants.TEMP_SOUND_PATH).getPath());
                        } catch (IOException e) {
                            e.printStackTrace();
                        }
                        mContext.startActivity(new Intent(mContext, ActivityRecordDub.class).putExtra(Constants.SOUND_PATH, Constants.TEMP_SOUND_PATH).putExtra(Constants.DUB_PATH, Constants.NOT_ASSIGNED));
                    }

                    @Override
                    public void failure(RetrofitError error) {
                        progressDialog.dismiss();
                        Toast.makeText(mContext, "خطا", Toast.LENGTH_SHORT).show();
                    }
                });
            }
        }
    }
}

File: app/src/main/kotlin/com.aspire.dubsmash/adapters/AdapterSounds.kt
package com.aspire.dubsmash.adapters

import android.app.Activity
import android.app.ProgressDialog
import android.media.MediaPlayer
import android.net.Uri
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageButton
import android.widget.RelativeLayout
import android.widget.TextView
import android.widget.Toast
import com.aspire.dubsmash.R
import com.aspire.dubsmash.fragments.OnFragmentInteractionListener
import com.aspire.dubsmash.util.*
import com.squareup.okhttp.Response
import retrofit.Callback
import retrofit.Retrofit
import java.io.FileOutputStream

/**
 * Created by sia on 11/20/15.
 */
class AdapterSounds(private val act: Activity, private var sounds: List<Sound>) : RecyclerView.Adapter<AdapterSounds.SoundsViewHolder>() {
    private var isPlaying: Boolean = false
    private var playingSoundButton: ImageButton? = null
    var mediaPlayer: MediaPlayer? = null
    private val callback: OnFragmentInteractionListener by lazy { act as OnFragmentInteractionListener }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): SoundsViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.item_sounds, parent, false)
        return SoundsViewHolder(view)
    }

    override fun onBindViewHolder(holder: SoundsViewHolder, position: Int) = holder.bindView(sounds[position])

    override fun getItemCount(): Int = sounds.size

    fun addData(data: List<Sound>) {
        sounds = data
        notifyDataSetChanged()
    }

    inner class SoundsViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView), View.OnClickListener {
        private val item: RelativeLayout by bindView(R.id.item)
        private val title: TextView by bindView(R.id.title)
        private val uploadedBy: TextView by bindView(R.id.uploaded_by)
        private val likeButton: ImageButton by bindView(R.id.like)
        private val playSound: ImageButton by bindView(R.id.play_sound)

        private val context = itemView.context
        private lateinit var sound: Sound

        init {
            setListeners()
        }

        fun bindView(sound: Sound) {
            title.text = sound.soundTitle
            if (sound.isLiked)
                likeButton.setImageResource(R.drawable.favorite_selected)
            else
                likeButton.setImageResource(R.drawable.favorite_unselected)
            uploadedBy.text = sound.soundSenderName
            this.sound = sound
        }

        private fun setListeners() {
            item.setOnClickListener(this)
            likeButton.setOnClickListener(this)
            playSound.setOnClickListener(this)
        }

        override fun onClick(v: View) {
            val id = v.id
            if (id == likeButton.id) {
                if (!sound.isLiked) {
                    likeButton.setImageResource(R.drawable.favorite_selected)
                    val result = networkApi.getSoundLikes(getUserId(context), sound.soundId)
                    result.enqueue(object : Callback<Response> {
                        override fun onResponse(response: retrofit.Response<Response>, retrofit: Retrofit?) {
                            throw UnsupportedOperationException()
                        }

                        override fun onFailure(t: Throwable?) {
                            throw UnsupportedOperationException()
                        }
                    })
                    sound.isLiked = true
                } else {
                    likeButton.setImageResource(R.drawable.favorite_unselected)
                    sound.isLiked = false
                }
            } else if (id == playSound.id) {
                if (isPlaying) {
                    stopAndReleaseMediaPlayer(mediaPlayer)
                    playingSoundButton?.setImageResource(R.drawable.ic_play)
                    playingSoundButton = playSound
                    playSound.setImageResource(R.drawable.download)
                } else {
                    playingSoundButton = playSound
                    playSound.setImageResource(R.drawable.download)
                    isPlaying = true
                }
                val result = networkApi.downloadSingleData(sound.soundUrl)
                result.enqueue(object : Callback<Response> {
                    override fun onResponse(response: retrofit.Response<Response>, retrofit: Retrofit?) {
                        FileOutputStream(tempSoundPath).write(response.body().body().bytes())
                        if (playingSoundButton === playSound) {
                            playSound.setImageResource(R.drawable.ic_pause)
                            mediaPlayer = MediaPlayer.create(context, Uri.parse(tempSoundPath))
                            mediaPlayer!!.start()
                            mediaPlayer!!.setOnCompletionListener {
                                playSound.setImageResource(R.drawable.ic_play)
                                mediaPlayer!!.reset()
                                mediaPlayer!!.release()
                                mediaPlayer = null
                                isPlaying = false
                            }
                        }
                    }

                    override fun onFailure(t: Throwable?) {
                        throw UnsupportedOperationException()
                    }

                })
            } else if (id == item.id) {
                val progressDialog = ProgressDialog(context)
                progressDialog.setMessage("کمی صبر کنید")
                progressDialog.setCancelable(false)
                progressDialog.show()
                val result = networkApi.downloadSingleData(sound.soundUrl)
                result.enqueue(object : Callback<Response> {
                    override fun onResponse(response: retrofit.Response<Response>, retrofit: Retrofit?) {
                        progressDialog.dismiss()
                        FileOutputStream(tempSoundPath).write(response.body().body().bytes())
                        //fixme arguments which should be sent to record dub fragment
                        callback.switchFragmentTo(FragmentId.RECORD_DUB)
                    }

                    override fun onFailure(t: Throwable?) {
                        progressDialog.dismiss()
                        Toast.makeText(context, "خطا", Toast.LENGTH_SHORT).show()
                    }
                })
            }
        }
    }
}
